### PR TITLE
pthread: implement cleanup_push/_pop

### PR DESCRIPTION
**Pull request by lukileczo**
fixes https://github.com/phoenix-rtos/phoenix-rtos-project/issues/827

DONE: RTOS-671

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [x] New test added: https://github.com/phoenix-rtos/phoenix-rtos-tests/pull/289
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: pthread/pthread.c [hunk 1]
```diff
@@ -183,6 +191,22 @@ static void pthread_full_key_cleanup(_pthread_key_cleanup_t *head)
...
 
+static void _pthread_do_cleanup(pthread_ctx *ctx)
+{
+	pthread_cleanup_t *head = ctx->cleanup_list;
+	while (ctx->cleanup_list != NULL) {
+		void (*routine)(void *) = head->routine;
+		void *arg = head->arg;
+		ctx->cleanup_list = head->next;
+		free(head);
+		head = ctx->cleanup_list;
```
Comment (agkaminski) [inline pthread/pthread.c:202]: imho better to do this after reacquiring the lock, cleanup_list head might've changed in the meantime

File: pthread/pthread.c [hunk 2]
```diff
@@ -183,6 +191,22 @@ static void pthread_full_key_cleanup(_pthread_key_cleanup_t *head)
 }
 
 
+static void _pthread_do_cleanup(pthread_ctx *ctx)
+{
+	pthread_cleanup_t *head = ctx->cleanup_list;
+	while (ctx->cleanup_list != NULL) {
```
Comment (agkaminski) [inline pthread/pthread.c:197]: Doesn't really matter, just seems more right to me, just a suggestion
```suggestion
	while (head != NULL) {
```

File: pthread/pthread.c [hunk 3]
```diff
@@ -1171,6 +1204,72 @@ void _pthread_atfork_child(void)
...
+		if (head == NULL) {
+			_pthread_ctx_put(ctx);
+			return;
+		}
+		head->next = ctx->cleanup_list;
+		ctx->cleanup_list = head;
+	}
+
+	ctx->cleanup_list->routine = routine;
+	ctx->cleanup_list->arg = arg;
```
Comment (agkaminski) [inline pthread/pthread.c:1236]: Somewhat shorter
```suggestion
	pthread_cleanup_t *head = (pthread_cleanup_t *)malloc(sizeof(pthread_cleanup_t));
	if (head == NULL) {
		_pthread_ctx_put(ctx);
		return;
	}
	
	head->next = ctx->cleanup_list;
	ctx->cleanup_list = head;

	ctx->cleanup_list->routine = routine;
	ctx->cleanup_list->arg = arg;
```

File: pthread/pthread.c [hunk 4]
```diff
@@ -1171,6 +1204,72 @@ void _pthread_atfork_child(void)
...
+
+	if (execute != 0) {
+		void (*routine)(void *) = head->routine;
+		void *arg = head->arg;
+		free(head);
+		routine(arg);
+	}
+	else {
+		free(head);
+	}
```
Comment (agkaminski) [inline pthread/pthread.c:1269]:
```suggestion
		routine(arg);
	}

	free(head);
```

Comment (agkaminski) [reply pthread/pthread.c:1269]: Or is it essential to `free(head)` before `routine()`? Can we e.g. die in the `routine()`?

Comment (lukileczo) [reply pthread/pthread.c:1269]: I'd leave it as it is, nothing in POSIX prevents us from calling e.g. `pthread_exit` from cleanup routine, so we'd end up with unfreed memory

**Review by agkaminski — APPROVED** [review]
LGTM

**Pull request merged**